Make `target_desired_distance` affect the navigation of `NavigationAgent2D/3D`

ershn opened the pull request:
Make `target_desired_distance` affect the navigation of `NavigationAgent2D/3D`.
- If the target is reachable, end the navigation only when the target is reached
- If the target is unreachable, end the navigation when the last waypoint is reached (same as before)

I created a test project to compare the old and new behavior.
[Navigation agent tests.zip](https://github.com/godotengine/godot/files/13059565/Navigation.agent.tests.zip)

Below are screenshots of running the test scenes on 4.1 and this branch.

<details>
<summary>Test scene screenshots</summary>

## `NavigationAgent2D`

### 4.1

![image](https://github.com/godotengine/godot/assets/7790167/df9e7fcb-16b4-4b77-bcbc-a7783ee3d9c7)
![image](https://github.com/godotengine/godot/assets/7790167/851a648c-7230-4b97-b48b-889a38827810)
![image](https://github.com/godotengine/godot/assets/7790167/b73d4ab9-40b5-4e06-a83d-ada7b623020d)
![image](https://github.com/godotengine/godot/assets/7790167/d4ae29ec-346c-4187-a7f9-98cd58d770d2)
![image](https://github.com/godotengine/godot/assets/7790167/f826ad2e-f0dc-455f-8ce2-86f5489a2fed)

### This branch

![image](https://github.com/godotengine/godot/assets/7790167/20103d7d-179b-4285-9955-3361597bfb41)
![image](https://github.com/godotengine/godot/assets/7790167/c20089f7-d897-42a3-ba44-c2dcf1b66fe1)
![image](https://github.com/godotengine/godot/assets/7790167/8b3b94a3-cbc9-4910-8a3e-c4704ab0be51)
![image](https://github.com/godotengine/godot/assets/7790167/410024d0-2fba-4d3f-8a69-8cc2908bda3c)
![image](https://github.com/godotengine/godot/assets/7790167/bb5dd1b7-0425-4446-bfb6-b159bc139d57)

## `NavigationAgent3D`

Legend
- red sphere -> waypoint area
- red cylinder -> position where a `waypoint_reached` signal was emitted
- blue sphere -> target area
- blue cylinder -> position where the `target_reached` signal was emitted
- white cylinder -> position where the `navigation_finished` signal was emitted

### 4.1

![image](https://github.com/godotengine/godot/assets/7790167/319adf80-7f41-489e-8d5d-acbd30023ca8)
![image](https://github.com/godotengine/godot/assets/7790167/87108419-6e35-41d0-9cde-fe06bc76eff6)
![image](https://github.com/godotengine/godot/assets/7790167/e78ee6f2-c0ec-414c-add6-d6d98f019674)
![image](https://github.com/godotengine/godot/assets/7790167/2ccde160-0be5-4dbf-a87a-3dd64aa2af90)
![image](https://github.com/godotengine/godot/assets/7790167/cda2f512-21a9-4ae1-bdb9-1d4876f843c9)

### This branch

![image](https://github.com/godotengine/godot/assets/7790167/2f663334-9dfc-4a4c-badd-128c409e237d)
![image](https://github.com/godotengine/godot/assets/7790167/81872fd8-9bde-4664-9135-11cad698f657)
![image](https://github.com/godotengine/godot/assets/7790167/6108d20c-1a5a-4cb8-8e64-91753f87df7a)
![image](https://github.com/godotengine/godot/assets/7790167/6562b991-37a7-44f4-9e52-2075a32f1c9c)
![image](https://github.com/godotengine/godot/assets/7790167/25645506-96f6-4487-9228-5d8b3fc27a16)

</details>

Notes
- The diff is bigger than I would have liked but a good chunk of it is because I extracted some code from `update_navigation` to individual functions (either out of necessity or to be able to express the algorithm with higher level methods). There isn't actually that much new code.
- I also updated the documentation of the signals/methods/properties that were affected by this change. Some of the changes are because the behavior changed, others are because the documentation wasn't actually 100% correct from the start.

- Fixes #82560
*Bugsquad edit:*
- Fixes https://github.com/godotengine/godot/issues/85247

ershn commented:
This is unrelated to the PR but I have some questions about the implementation of NavigationAgents.
- Is there a reason NavigationAgent2D/3D are Nodes and not Node2Ds/Node3Ds ? IMO making them Node2D/3Ds will remove the need to set a parent and allow to get the position directly in the node. It should then simplify the implementation / remove the possibility for usage errors.
- Currently the state update happens through calls to `get_next_path_position`/`is_navigation_finished`/`get_final_position` and to some degree when a NOTIFICATION_INTERNAL_PHYSICS_PROCESS is received. This way you only pay for what you use but this can lead to some finicky behavior. The same update function can also end up running several times a frame. What about updating everything on a NOTIFICATION_INTERNAL_PHYSICS_PROCESS and making the above methods `const` ? The API doesn't really need to change. Should I open a proposal for that ?

smix8 commented:
I think what should change is to make sure that the `target_desired_distance` is never allowed to be lower than the `path_desired_distance`. The NavigationAgent defaults should be updated to match this. Maybe print a warning if a user tries to set the value lower but clamp it anyway.

Users can set a `target_position` that is impossible because it is never reachable, e.g. because there is no connected navigation mesh in reach of that position. The current changes in this pr would make the distance check run forever every physics tick for an impossible check or for a path that already has ended and reset state. That should imo not be applied like that. 

@ershn 
You could join the navigation channel in the Godot Contributors Chat (https://chat.godotengine.org/) if you have general questions about the navigation system so we don't derail this pr. The short answer to your questions is physics / node / legacy jank.

ershn commented:
> I think what should change is to make sure that the target_desired_distance is never allowed to be lower than the path_desired_distance. The path following is stopped and the state reset when the path end is reached, there shouldn't be random checks running after this point. The NavigationAgent defaults should be updated to match this. Maybe print a warning if a user tries to set the value lower but clamp it anyway.

Yes, that's a reasonable fix. But aren't there genuine use cases where you would want the agent to loosely follow the path but still end up close to the target ? I honestly don't need that in my project right now but that doesn't seem to be an unreasonable requirement.
I just thought of another way to fix the problem. In case the target is reachable, for the last waypoint distance check use `target_desired_distance` instead of `path_desired_distance`. This way we can have `target_desired_distance` to be lower than `path_desired_distance` without having to run a check after the navigation is finished. It also doesn't seem to contradict the class documentation. What do you think ?

> You could join the navigation channel in the Godot Contributors Chat (https://chat.godotengine.org/) if you have general questions about the navigation system so we don't derail this pr.

Sure ! I didn't know about that chat. I'll ask my questions there then. Thank you !

smix8 commented:
> But aren't there genuine use cases where you would want the agent to loosely follow the path but still end up close to the target ?

Yes there could. To support this the pathfollowing would need to be changed so that it does not reset immediately on the last path array point. So this would need to change the check in a way similar like you already outlined. E.g. use the target_desired_distance if feasible but use the path_desired_distance still if the target is not right or reachable.

ershn commented:
Ok, I'll update the PR to do that then.

ershn committed:
Make target_desired_distance affect the navigation of NavigationAgent2D/3D

When the target is reachable, stop the navigation only when the target is reached.

ershn commented:
I updated the PR to implement the behavior I described in a previous comment.

> In case the target is reachable, for the last waypoint distance check use target_desired_distance instead of path_desired_distance.

Please note that this will slightly change the timing at which the `waypoint_reached` and `navigation_finished` signals are fired (maybe `link_reached` too ?). Until now, in case the target was reachable, for the last waypoint these signals were fired when the agent's position changed to be within `path_desired_distance` to `target_position`. Now they will be fired when the agent's position changes to be within `target_desired_distance` to `target_position`.
I think that change is ok because the documentation said nothing about this and the new behavior makes sense to me.

Additional notes:
- I also updated the class reference to add more details and make it read a bit better. Also changed `though` to `but` to make it sound more professional.
- I had to do some refactoring to avoid code duplication / having update_navigation call itself.
- I took the opportunity to rename `update_navigation` to `_update_navigation` to stick with private methods' naming convention.

Also, I'm not sure if we still need to call `_check_distance_to_target` in `NOTIFICATION_INTERNAL_PHYSICS_PROCESS` since we are now using `target_desired_distance` for distance checks directly. It's probably necessary if the agent teleports to the target_position without following the path but do we have to account for that ?

ershn commented:
I pushed a small fix to the class reference. I was confusing `e.g.` with `i.e.`

ershn commented:
Thinking about the problem again, I think the fix in its current form would lead to surprising behavior in some specific cases.

Until now, `target_desired_distance` didn't affect the navigation. It just controlled when `is_target_reached` changes to `true`, when `target_reached` is fired, and the value of `is_target_reachable`. This fix would make it affect the navigation too, and therefore make reasoning about `NavigationAgents` more difficult. IMO the effect `target_desired_distance` has on the agent shouldn't change with this PR (target distance related methods/properties are actually pretty independent from the rest of navigation agents, they are just convenience methods you could easily implement in user scripts).

Also, in case `target_desired_distance` is large enough that it encompasses a waypoint with its `path_desired_distance` radius, the timing at which `navigation_finished` is fired would be a bit surprising.

The diagram below shows how the navigation agent behaves currently.

![image](https://github.com/godotengine/godot/assets/7790167/379db7b6-68a4-40ea-8fd1-0782865f52cf)

- red circle: path desired distance
- blue circle: target desired distance
- orange line: path taken by the agent
- grey line: path taken by the agent if it doesn't stop on `navigation_finished`
- blue dot: `target_reached` signal + `is_target_reached` turns to `true`
- green dot: `waypoint_reached` signal
- purple dot: `navigation_finished` signal + `is_navigation_finished` turns to `true` + last `waypoint_reached` signal

Below is how the behavior would change with this fix.

![image](https://github.com/godotengine/godot/assets/7790167/250d8519-f12e-476e-9d8e-24d055ce11fe)

Conceptually the current fix would make it so that `navigation_finished` signals when we reached the target. However as you can see in the diagram above, this wouldn't always hold true. In the above scenario, I also don't think that stopping the navigation when the blue circle is reached is the correct solution.

@smix8 I think now that we should just make it so that the target distance check runs until necessary, i.e. when the target is reachable, keep running the target distance check on `NOTIFICATION_INTERNAL_PHYSICS_PROCESS` until the target is reached. I cannot think of other easy ways to fix the problem. What do you think ?

---

For reference, the diagram below illustrates the problem I'm trying to fix with this PR.

![image](https://github.com/godotengine/godot/assets/7790167/74a90c9d-5829-4903-802e-ccc0bcdeeaa1)

As you can see there is no blue dot. The `target_reached` signal is never fired and `is_target_reached` never turns to `true`.

smix8 commented:
> I also don't think that stopping the navigation when the blue circle is reached is the correct solution.

I think it would be. The `target_desired_distance` was designed to control when the pathfollowing should end at the last path position and override the `path_desired_distance` behavior for that position. That is the entire reason why this was separated into its own property.

ershn commented:
I see. If that's the intended effect of `target_desired_distance`, the current behavior definitely needs a fix.

Before making changes, I'd like your opinion on the following scenario. What do you think should happen here ?

![image](https://github.com/godotengine/godot/assets/7790167/456af7a9-fc47-449e-9528-4356adb4cc39)

If we end the navigation when the blue circle is reached, the last returned waypoint will actually be outside of the target's radius. But maybe there isn't much we can do here ? (and we should let the user be responsible for stopping the movement when the target is reached ?)

smix8 commented:
> If we end the navigation when the blue circle is reached, the last returned waypoint will actually be outside of the target's radius. But maybe there isn't much we can do here ?

For that to even happen users would need to set the `target_desired_distance` willingly to such a large value so the result is on them.

ershn commented:
I agree that users are responsible for the settings they use and that common cases should be prioritized.
But what I was trying to make sure with that example is that the algorithm is sound, i.e. it makes sense in all cases. IMO testing an algorithm at its limits is a good way to check if it's actually correct (that's how I found the bug above).

We seem to agree that there is no problem here, so I'll proceed with the changes 👍

ershn commented on scene/2d/navigation_agent_2d.h:
```diff
@@ -232,9 +232,12 @@ class NavigationAgent2D : public Node {
+	void _update_navigation();
 	void _request_repath();
 	void _check_distance_to_target();
+	bool _is_waypoint_reached(Vector2 origin);
```
Fixed

ershn commented on scene/3d/navigation_agent_3d.h:
```diff
@@ -245,9 +245,12 @@ class NavigationAgent3D : public Node {
+	void _update_navigation();
 	void _request_repath();
 	void _check_distance_to_target();
+	bool _is_waypoint_reached(Vector3 origin);
```
Fixed

ershn commented on scene/3d/navigation_agent_3d.cpp:
```diff
@@ -831,6 +839,20 @@ void NavigationAgent3D::_check_distance_to_target() {
 	}
 }
 
+bool NavigationAgent3D::_is_waypoint_reached(Vector3 origin) {
```
Fixed

ershn commented on scene/2d/navigation_agent_2d.cpp:
```diff
@@ -785,6 +793,19 @@ void NavigationAgent2D::_check_distance_to_target() {
 	}
 }
 
+bool NavigationAgent2D::_is_waypoint_reached(Vector2 origin) {
```
Fixed

ershn commented on doc/classes/NavigationAgent2D.xml:
```diff
@@ -180,7 +180,7 @@
 			The distance to search for other agents.
 		</member>
 		<member name="path_desired_distance" type="float" setter="set_path_desired_distance" getter="get_path_desired_distance" default="20.0">
-			The distance threshold before a path point is considered to be reached. This allows agents to not have to hit a path point on the path exactly, but only to reach its general area. If this value is set too high, the NavigationAgent will skip points on the path, which can lead to leaving the navigation mesh. If this value is set too low, the NavigationAgent will be stuck in a repath loop because ...
```
As far as I understand undershooting is not a problem because the agent will simply move a little bit closer to the waypoint each frame.

ershn commented on scene/2d/navigation_agent_2d.cpp:
```diff
@@ -274,7 +274,6 @@ void NavigationAgent2D::_notification(int p_what) {
 						NavigationServer2D::get_singleton()->agent_set_velocity_forced(agent, velocity_forced);
 					}
 				}
-				_check_distance_to_target();
```
Since reaching the target now ends the navigation, it's not enough anymore to just try to update `target_reached` and emit the corresponding signal on each frame. We need to also update the navigation state and I think this should only be done in `_update_navigation`.

ershn replied:
Since reaching the target now ends the navigation, it's not enough anymore to just try to update `target_reached` and emit the corresponding signal on each frame. We need to also update the navigation state and I think this should only be done in `_update_navigation`.

ershn commented on scene/2d/navigation_agent_2d.cpp:
```diff
@@ -687,102 +695,145 @@ void NavigationAgent2D::update_navigation() {
 	}
+
+	emit_signal(SNAME("navigation_finished"));
+}
```
`_trigger_waypoint_reached` and `_transition_to_navigation_finished` only contain code that was previously in `update_navigation`. The code is unchanged.

ershn commented on doc/classes/NavigationAgent2D.xml:
```diff
@@ -180,7 +180,7 @@
 		</member>
 		<member name="path_desired_distance" type="float" setter="set_path_desired_distance" getter="get_path_desired_distance" default="20.0">
-			The distance threshold before a path point is considered to be reached. This allows agents to not have to hit a path point on the path exactly, but only to reach its general area. If this value is set too high, the NavigationAgent will skip points on the path, which can lead to leaving the navigation mesh. If this value is set too low, the NavigationAgent will be stuck in a repath loop because ...
+			The distance threshold before a path point is considered to be reached. This allows agents to not have to hit a path point on the path exactly, but only to reach its general area. If this value is set too high, the NavigationAgent will skip points on the path, which can lead to it leaving the navigation mesh. If this value is set too low, the NavigationAgent will be stuck in a repath loop beca ...
```
As far as I understand undershooting is not a problem because the agent will simply move a little bit closer to the waypoint each frame.

ershn commented:
This proved to be more work than I expected but the PR should be ready for review now.
I also updated the PR description with test scene screenshots.

TitanShadow12 commented:
> I see ! If that's the intended effect of `target_desired_distance`, the current behavior definitely needs a fix then.
> 
> Before making changes, I'd like your opinion on the following scenario. What do you think should happen here ?
> 
> ![image](https://user-images.githubusercontent.com/7790167/273438651-456af7a9-fc47-449e-9528-4356adb4cc39.png)
> 
> If we end the navigation when the blue circle is reached, the last returned waypoint will actually be outside of the target's radius. But maybe there isn't much we can do here ? (and we should let the user be responsible for stopping the movement when the target is reached ?)

I originally thought navigation should finish once the destination waypoint is reached:
- path to all waypoints using path_desired_distance
- if the next waypoint is the last one, then consider it reached if within target_desired_distance instead of path_desired_distance

I like this implementation though, since it allows you to preempt the pathfinding if you get within target_desired_distance (if that is greater than path_desired_distance) - possibly allowing an agent to (maybe only incidentally) avoid pathing around a wall if unnecessary.

I'm fine with this fix because it solves my particular problem - I don't want my agents to be picky about hitting their path waypoints, but I want them to stop almost exactly where they're told. The 4.1 implementation doesn't take this difference into account, and I'm excited to see this change.

ershn commented:
Rebased and re-tested.

smix8 commented:
I tested it for the discussed changes and it works in general fine for the target desired distance as I would expect but what worked unexpectedly is that target desired distance seem to not reliably or at all overrule the path desired distance for the last path position.

E.g. if I set a NavigationAgent `target_desired_distance` to 0.5 and the `path_desired_distance` to 5.0 the agent does never move in to 0.5 distance of the target. It stops the path at 5.0 distance to the last path position. Since in my test I immediately query a new path to the next patrol position when the path ends my agent basically never even comes close to the target position.

ershn commented:
@smix8 I can't reproduce the problem you're seeing on my side. Could you provide a test project to reproduce the behavior you are seeing ? Or maybe the source code you are using for the agent ?

> if I set a NavigationAgent target_desired_distance to 0.5 and the path_desired_distance to 5.0 the agent does never move in to 0.5 distance of the target. It stops the path at 5.0 distance to the last path position.

The only reason I can see this happening is because the target is unreachable. If the target is unreachable the `target_desired_distance` doesn't override `path_desired_distance`.

Or maybe it's because you are requesting a new path directly when reaching the last waypoint ? E.g. even after this change and even if `target_desired_distance` < `path_desired_distance`, `waypoint_reached` will be sent when the agent moves within `path_desired_distance` of the last waypoint. If you want the agent to reach the target you need to wait for the `target_reached` or `navigation_finished` signal.

The test project I uploaded in the description tests for the scenario you are describing but everything seems to be working as expected. In the images below, the white cross/cylinder is also the final position of the agent.

![277079342-8b3b94a3-cbc9-4910-8a3e-c4704ab0be51](https://github.com/godotengine/godot/assets/7790167/04b34a5d-b708-486e-a564-b894e38f01c5)
![277079392-6108d20c-1a5a-4cb8-8e64-91753f87df7a](https://github.com/godotengine/godot/assets/7790167/5fb6bc0a-67eb-4a53-8e4f-8fa9a198dcbc)

smix8 commented:
I tested the 3D scene with path_desired_distance = 2.0 and target_desired_distance = 0.1.

![path](https://github.com/godotengine/godot/assets/52464204/21930c0d-a61f-47cd-b23f-7c8587e2c0ff)

The agent stops at the path desired distance circle.

I noticed that the problem happens with any target desired distance value below 0.5 in that scene. If the value is >= 0.5 it works as expected.

ershn commented:
Ah I see ! In that scene `0.5` is the y distance between the scene geometry and the computed navigation mesh.
The problem was that `get_final_position` didn't subtract `path_height_offset` from the position it returned. This position was then used by `is_target_reachable` which resulted in the target being considered unreachable when `target_desired_distance < 0.5`. This bug should be fixed now.
It seems this problem has been there for a while but it became more apparent with this PR, since `is_target_reachable` now affects navigation.

On an unrelated note, I took this opportunity to also change the order in which the `target_reached` and `navigation_finished` signals are sent (`target_reached` first now). This will allow using `is_target_reached` in `navigation_finished` signal callbacks. Re-tested everything after this change.

smix8 commented:
I tested the change with the 3D demo scene but with any target_desired_distance below 0.5 is enough to get the agent stuck on the last path_desired_distance red circle instead of moving closer to the target_desired_distance blue circle.

ershn commented:
Sorry but I really can't reproduce the problem this time. Could you provide more info ?

I did a fresh build on this branch, downloaded the zip in the description, and run every 3D scene with values of `0.1` (the minimum), `0.4`, `0.5` for `target_desired_distance`. I also tested a bunch of different combinations with `path_desired_distance` but I couldn't find a case where it doesn't work as expected.

It also doesn't seem like I forgot to push a commit.

![image](https://github.com/godotengine/godot/assets/7790167/ecd8a8aa-1fc1-4592-9035-66e8019479a1)
![image](https://github.com/godotengine/godot/assets/7790167/bec9ed8d-7461-497d-91a7-9797778ebefd)

Please note that you need to have a `path_height_offset` of `0.5` (the default) in the 3D test scenes, otherwise the target won't be considered reachable.

smix8 commented:
> Sorry but I really can't reproduce the problem this time. 

It is fine and works. I noticed I made a mistake. I set some properties in script from testing. When I started to print those values I noticed that they did not match what I set on the NavigationAgent node.

ershn commented on scene/2d/navigation_agent_2d.cpp:
```diff
@@ -687,102 +695,145 @@ void NavigationAgent2D::update_navigation() {
-		const Vector<int32_t> &navigation_path_types = navigation_result->get_path_types();
-		const TypedArray<RID> &navigation_path_rids = navigation_result->get_path_rids();
-		const Vector<int64_t> &navigation_path_owners = navigation_result->get_path_owner_ids();
+	// Check if the navigation has already finished
```
Fixed all the comments.

oxeron commented:
When can we hope a release ? I think I'm affected by this bug

YuriSizov reviewed (approved):
Looks great to me! Documentation changes alone are very much appreciated, and improved logic and code organization are just a cherry on top 🙂

YuriSizov commented:
Thanks!

PythonCircuit commented:
> When can we hope a release ? I think I'm affected by this bug

I am trying to create an agent that utilizes Path Desired Distance with .1 and Target Desired Distance with .1 but it just seems to get stuck in a re-loop 99.99% of the time, the goal is to have 99.99% precision in my case . Hopefully this fixes that issue if I understand this thread correctly.  I have not tested or reviewed the code being implemented here. 
I have read documentation and example code  possible for setting up a simple navigation region 3d on simple geometry with basic settings with character controlled agent that simply clicks a position on a map and uses the navigation system to get there. I also cannot get the navigation mesh to properly work with default settings on very smooth geometry for hills inclines/slopes without setting the cell height to .01-.1f which seems to cause other issues I cannot fathom. I personally am using Terrain3D for terrain but had exact same issues on a default plane/block

oxeron commented:
Any news ?
